jinahub/rankers/CatboostRanker/catboost_ranker.py

```diff
@@ -0,0 +1,168 @@
+
+import numpy as np
+from catboost import CatBoostRanker, Pool
+
+from jina import DocumentArray, Executor, requests
+from jina.logging.logger import JinaLogger
+
+
+class CatboostRanker(Executor):
+    def __init__(
+        self,
+        query_features: List[str],
+        match_features: List[str],
+        label: str,
+        weight: Optional[str] = None,
+        model_path: Optional[str] = None,
+        catboost_parameters: Dict = {
+            'custom_metric': ['NDCG', 'AverageGain:top=10'],
+            'verbose': False,
+            'random_seed': 0,
+            'loss_function': 'RMSE',
+            'num_trees': 5,
+        },
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.q_features = query_features
+        self.m_features = match_features
+        self.label = label
+        self.weight = weight
+        self.model_path = model_path
+        self.catboost_parameters = catboost_parameters
+        self.model = None
+        self.logger = JinaLogger('catboost logger')
+        if self.model_path and os.path.exists(self.model_path):
+            self._load_model()
+        else:
+            self.model = CatBoostRanker()
+
+    def _load_model(self, path):
+        """Load model from model path"""
+        self.model = CatBoostRanker().load_model(path)
+
+    def _save_model(self, path):
+        """Dump model into cbm."""
+        self.model.save_model(fname=path, format='cbm')
+
+    def _extract_features(self, docs: DocumentArray):
+        """Loop over docs, build dataset to train the model.
+
+        This func get all :attr:`query_feature` from tags inside :class:`Document`,
+        and get all :attr:`match_features` from all matched :class:`Document` for each query document.
+        Each query document corresponded to N matched documents, and each query-document pair features
+        are combined into a feature vector. We stack all feature vectors and return the training data.
+        """
+        group_ids = []
+        label_vector = []
+        feature_vectors = []
+        for idx, doc in enumerate(docs):
+            q_feature_vector = [
+                doc.tags.get(query_feature) for query_feature in self.q_features
+            ]
+            for match in doc.matches:
+                group_ids.append(idx)
+                label_vector.append(match.tags.get(self.label))
+                m_feature_vector = [
+                    match.tags.get(docum_feature) for docum_feature in self.m_features
+                ]
+                feature_vectors.append(q_feature_vector + m_feature_vector)
+        return np.array(feature_vectors), label_vector, group_ids
+
+    def _extract_weights(self, docs: DocumentArray):
+        """Extract weights data from docs, weights are needed if the query_docs has different
+        importance. Once given, we'll get values stored in the Document.tags based on the field name.
+        """
+        weight_vector = []
```

Comment:
can we do it in a list-comprehension way?

Reply:
i don't see it brings anything, only reduce readability